### ISSUE TITLE
[v2] Merge state.go into raft.go?

### ISSUE THREAD

**Issue by ongardie** [hashicorp/raft#172]
I'll volunteer to start this off.

**Comment by ongardie-sfdc** [comment]
I'll hold off on physically moving code still, but I'd like to end up with:
- raft.go consisting of the public-facing API exported by the library (mostly in api.go today)
- server.go consisting of the `raftServer` struct and its methods (mostly in raft.go today)
- No state.go (folded into `raftServer`)
- No other files accessing `Raft` or `raftServer` (fsm.go, observer.go, snapshot.go, state.go do this today)

A few questions:

The functions `BootstrapCluster()`, `RecoverCluster()`, and `HasExistingState()` are a little unique, in that they're part of the public API but are most similar to code in `raftServer`. Should we move them to a new file like "offline.go"?

What is observer.go for? If it's just for unit tests, can we hide it from the public API?

One ugly bit right now is that a server can be a Follower, Candidate, Leader, or Shutdown, and shutting down sets the state.go state of the server to Shutdown forcefully from api.go. Any objection to removing Shutdown from the RaftState enum? A server that shut down would just stay in the Follower state. This would affect and possibly break some applications that call Raft.State() and compare it to things w.r.t. Shutdown.

Should we remove State(), LastIndex(), LastContact() from the API? Clients can get this information from Stats() instead. Alternatively, we could rewrite them in terms of Stats(). And what about Leader()? It's not in Stats today but probably should be.

@slackpad @superfell

**Comment by slackpad** [comment]
> The functions BootstrapCluster(), RecoverCluster(), and HasExistingState() are a little unique, in that they're part of the public API but are most similar to code in raftServer. Should we move them to a new file like "offline.go"?

I like offline.go for that - those are special and it's nice to make them clear they aren't for a running cluster.

> What is observer.go for? If it's just for unit tests, can we hide it from the public API?

This is only for unit tests so let's definitely hide it. We also don't want external users waiting on leader changes using this, as they are probably doing something racy and unreliable. Given how state has been reorganized we should also see if we can get rid of this - it's mostly used in unit tests to wait for things so settle.

> One ugly bit right now is that a server can be a Follower, Candidate, Leader, or Shutdown, and shutting down sets the state.go state of the server to Shutdown forcefully from api.go. Any objection to removing Shutdown from the RaftState enum?

This sounds ok - I don't know of any places that look at the state for shutdown in our stuff. What's the reason not to have this as a special state?

> Should we remove State(), LastIndex(), LastContact() from the API? Clients can get this information from Stats() instead. Alternatively, we could rewrite them in terms of Stats(). And what about Leader()? It's not in Stats today but probably should be.

Having one `Stats()` is probably better, esp. since you get an single snapshot of everything together. Adding the leader to `Stats()` makes sense and I can adjust that on the Consul side which I know uses `Leader()` now.

**Comment by ongardie-sfdc** [comment]
Thanks @slackpad.

> What's the reason not to have this as a special state?

My main issue with Shutdown is that api.go sets the state to Shutdown, rather than raftServer doing it in response to a channel. If we fix that, I'm not sure that we need Shutdown as a special state, but either way works for me.

**Comment by slackpad** [comment]
>  If we fix that, I'm not sure that we need Shutdown as a special state, but either way works for me.

That makes sense. Given that this has a future to wait for the shutdown it seems like we can drop the state. Consul doesn't make use of it.

**Comment by ongardie-sfdc** [comment]
With those last commits, not much of this original issue is left. I created #198 to track the remaining stuff, and I'm closing this one.

**Comment by ongardie-sfdc** [comment]
Oh just kidding, I don't have permission to close this. @slackpad, a little help? :)
